File: src/main/js/g0401_0500/s0494_target_sum/solution.js
// #Medium #Array #Dynamic_Programming #Backtracking #Big_O_Time_O(n*(sum+s))_Space_O(n*(sum+s))
// #2024_12_24_Time_2_ms_(99.66%)_Space_51.5_MB_(61.90%)

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var findTargetSumWays = function(nums, target) {
    const n = nums.length
    let totalSum = 0
    for (let i = 0; i < n; i++) {
        totalSum += nums[i]
    }

    const sum = totalSum - target
    if (sum < 0 || sum % 2 === 1) {
        return 0
    }

    const targetSum = Math.floor(sum / 2)

    // Helper function to solve the problem
    const solve = (nums, target) => {
        let prev = new Array(target + 1).fill(0)

        if (nums[0] === 0) {
            prev[0] = 2 // Two ways to form sum 0: +0 and -0
        } else {
            prev[0] = 1
        }

        if (nums[0] !== 0 && nums[0] <= target) {
            prev[nums[0]] = 1
        }

        for (let i = 1; i < nums.length; i++) {
            const curr = new Array(target + 1).fill(0)
            for (let j = 0; j <= target; j++) {
                const taken = j >= nums[i] ? prev[j - nums[i]] : 0
                const nonTaken = prev[j]
                curr[j] = taken + nonTaken
            }
            prev = curr
        }

        return prev[target]
    }

    return solve(nums, targetSum)
};

export { findTargetSumWays }

File: src/main/erlang/g0401_0500/s0494_target_sum/Solution.erl
% #Medium #Array #Dynamic_Programming #Backtracking #Big_O_Time_O(n*(sum+s))_Space_O(n*(sum+s))
% #2025_01_22_Time_377_(100.00%)_Space_60.11_(100.00%)

-spec find_target_sum_ways(Nums :: [integer()], Target :: integer()) -> integer().
find_target_sum_ways(Nums, Target) ->
    TotalSum = lists:sum(Nums),
    Sum = TotalSum - Target,
    case Sum < 0 orelse Sum rem 2 =:= 1 of
        true -> 0;
        false -> solve(Nums, Sum div 2)
    end.

-spec solve(Nums :: [integer()], Target :: integer()) -> integer().
solve([First|_] = Nums, Target) ->
    Prev = lists:duplicate(Target + 1, 0),
    InitPrev = case First of
        0 -> lists:sublist([2|lists:duplicate(Target, 0)], Target + 1);
        _ when First =< Target -> 
            lists:sublist([1] ++ lists:duplicate(First - 1, 0) ++ [1] ++ 
                         lists:duplicate(Target - First, 0), Target + 1);
        _ -> [1|lists:duplicate(Target, 0)]
    end,
    process_nums(tl(Nums), Target, InitPrev).

process_nums([], _Target, Prev) ->
    lists:last(Prev);
process_nums([Num|Rest], Target, Prev) ->
    Curr = lists:foldl(fun(J, Acc) ->
        Taken = case J >= Num of
            true -> lists:nth(J - Num + 1, Prev);
            false -> 0
        end,
        NonTaken = lists:nth(J + 1, Prev),
        Acc ++ [Taken + NonTaken]
    end, [], lists:seq(0, Target)),
    process_nums(Rest, Target, Curr).
